packages/create-turbo/src/commands/create/createProject.ts

```diff
@@ -0,0 +1,231 @@
+export class DownloadError extends Error {}
+
+export async function createProject({
+  appPath,
+  projectName,
+  example,
+  examplePath,
+}: {
+  appPath: string;
+  projectName: string;
+  example: string;
+  examplePath?: string;
+}): Promise<{
+  cdPath: string;
+  hasPackageJson: boolean;
+  availableScripts: Array<string>;
+}> {
+  let repoInfo: RepoInfo | undefined;
+  let repoUrl: URL | undefined;
+  const isDefaultExample = example === "basic" || example === "default";
+
+  try {
+    repoUrl = new URL(example);
+  } catch (err: any) {
+    if (err.code !== "ERR_INVALID_URL") {
+      error(err);
+      process.exit(1);
+    }
+  }
+
+  if (repoUrl) {
+    if (repoUrl.origin !== "https://github.com") {
+      error(
+        `Invalid URL: ${chalk.red(
+          `"${example}"`
+        )}. Only GitHub repositories are supported. Please use a GitHub URL and try again.`
+      );
+      process.exit(1);
+    }
+
+    repoInfo = await getRepoInfo(repoUrl, examplePath);
+
+    if (!repoInfo) {
+      error(
+        `Found invalid GitHub URL: ${chalk.red(
+          `"${example}"`
+        )}. Please fix the URL and try again.`
+      );
+      process.exit(1);
+    }
+
+    const found = await hasRepo(repoInfo);
+
+    if (!found) {
+      error(
+        `Could not locate the repository for ${chalk.red(
+          `"${example}"`
+        )}. Please check that the repository exists and try again.`
+      );
+      process.exit(1);
+    }
+  } else {
+    const found = await existsInRepo(example);
+
+    if (!found) {
+      error(
+        `Could not locate an example named ${chalk.red(
+          `"${example}"`
+        )}. It could be due to the following:\n`,
+        `1. Your spelling of example ${chalk.red(
+          `"${example}"`
+        )} might be incorrect.\n`,
+        `2. You might not be connected to the internet or you are behind a proxy.`
+      );
+      process.exit(1);
+    }
+  }
+
+  const root = path.resolve(appPath);
+
+  if (!(await isWriteable(path.dirname(root)))) {
+    error(
+      "The application path is not writable, please check folder permissions and try again."
+    );
+    error("It is likely you do not have write permissions for this folder.");
+    process.exit(1);
+  }
+
+  const appName = path.basename(root);
+  try {
+    await fs.mkdir(root, { recursive: true });
+  } catch (err) {
+    error("Unable to create project directory");
+    console.error(err);
+    process.exit(1);
+  }
+  const { isEmpty, conflicts } = isFolderEmpty(root);
+  if (!isEmpty) {
+    error(
+      `${chalk.dim(root)} has ${conflicts.length} conflicting ${
+        conflicts.length === 1 ? "file" : "files"
+      } - please try a different location`
+    );
+    process.exit(1);
+  }
+
+  const originalDirectory = process.cwd();
+  process.chdir(root);
+
+  /**
+   * clone the example repository
+   */
+  const loader = turboLoader("Downloading files...");
+  try {
+    if (repoInfo) {
+      const repoInfo2 = repoInfo;
```

Comment:
What is the motivation here? They are references to the same object no?

Reply:
This was lifted right from create-next-app, but I'll clean it up a bit